src/formula/variant.cpp

```diff
@@ -410,108 +315,93 @@ variant variant::operator[](const variant& v) const
 
 variant variant::get_keys() const
 {
-	must_be(TYPE_MAP);
-	assert(map_);
+	must_be(VARIANT_TYPE::TYPE_MAP);
+
 	std::vector<variant> tmp;
-	for(std::map<variant,variant>::const_iterator i=map_->begin(); i != map_->end(); ++i) {
-			tmp.push_back(i->first);
+	for(const auto& i : value_cast<game_logic::variant_map>()->get_container()) {
+		tmp.push_back(i.first);
 	}
+
 	return variant(&tmp);
 }
 
 variant variant::get_values() const
 {
-	must_be(TYPE_MAP);
-	assert(map_);
+	must_be(VARIANT_TYPE::TYPE_MAP);
+
 	std::vector<variant> tmp;
-	for(std::map<variant,variant>::const_iterator i=map_->begin(); i != map_->end(); ++i) {
-			tmp.push_back(i->second);
+	for(const auto& i : value_cast<game_logic::variant_map>()->get_container()) {
+		tmp.push_back(i.second);
 	}
+
 	return variant(&tmp);
 }
 
 variant_iterator variant::begin() const
 {
-	if(type_ == TYPE_LIST)
-		return variant_iterator( list_->begin() );
+	if(is_list()) {
+		return variant_iterator(value_cast<game_logic::variant_list>()->get_container().begin());
+	}
 
-	if(type_ == TYPE_MAP)
-		return variant_iterator( map_->begin() );
+	if(is_map()) {
+		return variant_iterator(value_cast<game_logic::variant_map>()->get_container().begin());
+	}
 
 	return variant_iterator();
 }
+
 variant_iterator variant::end() const
 {
-	if(type_ == TYPE_LIST)
-		return variant_iterator( list_->end() );
+	if(is_list()) {
+		return variant_iterator(value_cast<game_logic::variant_list>()->get_container().end());
+	}
 
-	if(type_ == TYPE_MAP)
-		return variant_iterator( map_->end() );
+	if(is_map()) {
+		return variant_iterator(value_cast<game_logic::variant_map>()->get_container().end());
+	}
 
 	return variant_iterator();
 }
 
 bool variant::is_empty() const
 {
-	if(type_ == TYPE_NULL) {
-		return true;
-	} else if (type_ == TYPE_LIST) {
-		assert(list_);
-		return list_->empty();
-	} else if (type_ == TYPE_MAP) {
-		assert(map_);
-		return map_->empty();
-	}
-
-	return false;
+	return value_->is_empty();
 }
 
 size_t variant::num_elements() const
 {
-	if(type_ == TYPE_CALLABLE) {
-		return 1;
-	}
-
-	if (type_ == TYPE_LIST) {
-		assert(list_);
-		return list_->size();
-	} else if (type_ == TYPE_MAP) {
-		assert(map_);
-		return map_->size();
-	} else {
-		throw type_error(formatter() << "type error: "
-			<< " expected a list or a map but found " << type_string()
-			<< " (" << to_debug_string() << ")");
```

Comment:
So do we still get this nice error message now? I get the impression that it's gone...